apps/aefate/src/aefa_fate_op.erl

```diff
@@ -630,7 +636,97 @@ str_join(Arg0, Arg1, Arg2, EngineState) ->
     write(Arg0, Result, ES3).
 
 str_length(Arg0, Arg1, EngineState) ->
-    un_op(str_length, {Arg0, Arg1}, EngineState).
+    {StrValue, ES1} = get_op_arg(Arg1, EngineState),
+    if not ?IS_FATE_STRING(StrValue) ->
+        aefa_fate:abort({value_does_not_match_type, StrValue, string}, ES1);
+       true -> ok
+    end,
+    case aefa_engine_state:vm_version(ES1) of
+        ?VM_FATE_SOPHIA_1 ->
+            Res = aeb_fate_data:make_integer(
+                    byte_size(?FATE_STRING_VALUE(StrValue))),
+            write(Arg0, Res, ES1);
+        Vm when Vm > ?VM_FATE_SOPHIA_1 ->
+            case unicode:characters_to_nfc_list(?FATE_STRING_VALUE(StrValue)) of
+                {error, _, _} ->
+                    aefa_fate:abort({value_does_not_match_type, StrValue, string_utf8}, ES1);
+                Chars ->
+                    Res = aeb_fate_data:make_integer(length(Chars)),
+                    ES2 = aefa_engine_state:spend_gas_for_new_cells(length(Chars), ES1),
+                    write(Arg0, Res, ES2)
+            end
+    end.
+
+str_to_list(Arg0, Arg1, EngineState) ->
+    {Value, ES1} = get_op_arg(Arg1, EngineState),
+    if not ?IS_FATE_STRING(Value) ->
+        aefa_fate:abort({value_does_not_match_type, Value, string}, ES1);
+       true -> ok
+    end,
+    case unicode:characters_to_nfc_list(?FATE_STRING_VALUE(Value)) of
+        {error, _, _} ->
+            aefa_fate:abort({value_does_not_match_type, Value, string_utf8}, ES1);
+        Chars ->
+            ES2 = aefa_engine_state:spend_gas_for_new_cells(length(Chars) * 2 + 1, ES1),
+            write(Arg0, Chars, ES2)
+    end.
+
+str_from_list(Arg0, Arg1, EngineState) ->
+    {Value, ES1} = get_op_arg(Arg1, EngineState),
+    if not ?IS_FATE_LIST(Value) ->
+        aefa_fate:abort({value_does_not_match_type, Value, string}, ES1);
+       true -> ok
+    end,
+    case unicode:characters_to_nfc_binary(Value) of
+        {error, _, _} -> aefa_fate:abort({value_does_not_match_type, Value, {list, char}}, ES1);
+        Str ->
+            Cells = string_cells(?FATE_STRING_VALUE(Str)),
+            ES2 = aefa_engine_state:spend_gas_for_new_cells(Cells + 1, ES1),
+            write(Arg0, ?FATE_STRING_VALUE(Str), ES2)
```

Comment:
Either use `aeb_fate_data:make_string/1` instead of `?FATE_STRING_VALUE`, or drop (since strings are untagged).

```erlang
?FATE_X_VALUE(TaggedX) -> UntaggedX
aeb_fate_data:make_X(UntaggedX) -> TaggedX
```